src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+      // from kzg_open, it could be done later or separate from computing W.
+
+      let compute_witness_polynomial = |f: &[E::Fr], u: E::Fr| -> Vec<E::Fr> {
+        let d = f.len();
+
+        // Compute h(x) = f(x)/(x - u)
+        let mut h = vec![E::Fr::ZERO; d];
+        for i in (1..d).rev() {
+          h[i - 1] = f[i] + h[i] * u;
+        }
+
+        h
+      };
+
+      let h = compute_witness_polynomial(f, u);
+
+      <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &h)
+        .comm
+        .preprocessed()
+    };
+
+    let kzg_open_batch = |C: &[E::G1Affine],
+                          f: &[Vec<E::Fr>],
+                          u: &[E::Fr],
+                          transcript: &mut <NE as NovaEngine>::TE|
+     -> (Vec<E::G1Affine>, Vec<Vec<E::Fr>>) {
+      let poly_eval = |f: &[E::Fr], u: E::Fr| -> E::Fr {
+        // Horner's scheme
+        let mut v = f[f.len() - 1];
+
+        for fi in f.iter().rev().skip(1) {
+          v *= u;
+          v += fi;
+        }
+
+        v
+      };
+
+      let scalar_vector_muladd = |a: &mut Vec<E::Fr>, v: &Vec<E::Fr>, s: E::Fr| {
+        assert!(a.len() >= v.len());
+        for i in 0..v.len() {
+          a[i] += s * v[i];
+        }
+      };
+
+      let kzg_compute_batch_polynomial = |f: &[Vec<E::Fr>], q: E::Fr| -> Vec<E::Fr> {
+        let k = f.len(); // Number of polynomials we're batching
+
+        let q_powers = Self::batch_challenge_powers(q, k);
+
+        // Compute B(x) = f[0] + q*f[1] + q^2 * f[2] + ... q^(k-1) * f[k-1]
+        let mut B = f[0].clone();
+        for i in 1..k {
+          scalar_vector_muladd(&mut B, &f[i], q_powers[i]); // B += q_powers[i] * f[i]
+        }
+
+        B
+      };
+      ///////// END kzg_open_batch closure helpers
+
+      let k = f.len();
+      let t = u.len();
+      assert!(C.len() == k);
+
+      // The verifier needs f_i(u_j), so we compute them here
+      // (V will compute B(u_j) itself)
+      let mut v = vec![vec!(E::Fr::ZERO; k); t];
+      for i in 0..t {
+        // for each point u
+        for (j, f_j) in f.iter().enumerate().take(k) {
+          // for each poly f
+          v[i][j] = poly_eval(f_j, u[i]); // = f_j(u_i)
+        }
+      }
+
+      let q = Self::get_batch_challenge(C, u, &v, transcript);
+      let B = kzg_compute_batch_polynomial(f, q);
+
+      // Now open B at u0, ..., u_{t-1}
+      let mut w = Vec::with_capacity(t);
+      for ui in u {
+        let wi = kzg_open(&B, *ui);
+        w.push(wi);
+      }
+
+      // Compute the commitment to the batched polynomial B(X)
+      let q_powers = Self::batch_challenge_powers(q, k);
+      let c_0: E::G1 = C[0].into();
+      let C_B = (c_0 + NE::GE::vartime_multiscalar_mul(&q_powers[1..k], &C[1..k])).preprocessed();
+
+      // The prover computes the challenge to keep the transcript in the same
+      // state as that of the verifier
+      let _d_0 = Self::verifier_second_challenge(&C_B, &w, transcript);
```

Comment:
No need to add `C_B` to the transcript since it is derived from `C` which is already in the transcript.
Moreover it is already computed by the verifier.

Reply:
Fixed in https://github.com/lurk-lab/arecibo/pull/206